po-docs/ja/guestmount.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Podwrapper::Man 1.60.1 (Pod::Simple 3.48)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "guestmount 1"
.TH guestmount 1 2026-07-14 libguestfs-1.60.1 "Virtualization Support"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH 名前
.IX Header "名前"
guestmount \- Mount a guest filesystem on the host using FUSE and libguestfs
.SH 書式
.IX Header "書式"
.Vb 1
\& guestmount [\-\-options] \-a disk.img \-m device [\-\-ro] mountpoint
\&
\& guestmount [\-\-options] \-a disk.img \-i [\-\-ro] mountpoint
\&
\& guestmount [\-\-options] \-d Guest \-i [\-\-ro] mountpoint
.Ve
.SH 説明
.IX Header "説明"
The guestmount program can be used to mount virtual machine filesystems and other disk images on the host.  It uses libguestfs for access to the guest filesystem, and FUSE (the "filesystem in userspace") to make it appear as a mountable device.
.PP
Along with other options, you have to give at least one device (\fI\-a\fR option) or libvirt domain (\fI\-d\fR option), and at least one mountpoint (\fI\-m\fR option) or use the \fI\-i\fR inspection option.  How this works is better explained in the \fBguestfish\fR\|(1) manual page, or by looking at the examples below.
.PP
FUSE lets you mount filesystems as non\-root.  The mountpoint must be owned by you.  The filesystem will not be visible to any other users unless you make configuration changes, see "NOTES" below.
.PP
To unmount the filesystem, use the \fBguestunmount\fR\|(1) command.
.SH 例
.IX Header "例"
For a typical Windows guest which has its main filesystem on the first partition:
.PP
.Vb 1
\& guestmount \-a windows.img \-m /dev/sda1 \-\-ro /mnt
.Ve
.PP
For a typical Linux guest which has a /boot filesystem on the first partition, and the root filesystem on a logical volume:
.PP
.Vb 1
\& guestmount \-a linux.img \-m /dev/VG/LV \-m /dev/sda1:/boot \-\-ro /mnt
.Ve
.PP
To get libguestfs to detect guest mountpoints for you:
.PP
.Vb 1
\& guestmount \-a guest.img \-i \-\-ro /mnt
.Ve
.PP
For a libvirt guest called "Guest" you could do:
.PP
.Vb 1
\& guestmount \-d Guest \-i \-\-ro /mnt
.Ve
.PP
If you don’t know what filesystems are contained in a guest or disk image, use \fBvirt\-filesystems\fR\|(1) first:
.PP
.Vb 1
\& virt\-filesystems \-d MyGuest
.Ve
.PP
If you want to trace the libguestfs calls but without excessive debugging information, we recommend:
.PP
.Vb 1
\& guestmount [...] \-\-trace /mnt
.Ve
.PP
If you want to debug the program, we recommend:
.PP
.Vb 1
\& guestmount [...] \-\-trace \-\-verbose /mnt
.Ve
.PP
To unmount the filesystem after using it:
.PP
.Vb 1
\& guestunmount /mnt
.Ve
.SH 注記
.IX Header "注記"
.SS 他のユーザーは標準でファイルシステムを参照できません。
.IX Subsection "他のユーザーは標準でファイルシステムを参照できません。"
If you mount a filesystem as one user (eg. root), then other users will not be able to see it by default.  The fix is to add the FUSE \f(CW\*(C`allow_other\*(C'\fR option when mounting:
.PP
.Vb 1
\& sudo guestmount [...] \-o allow_other /mnt
.Ve
.PP
\&\fBand\fR to enable this option in \fI/etc/fuse.conf\fR.
.SS "Enabling FUSE"
.IX Subsection "Enabling FUSE"
On some distros, you may need to add yourself to a special group (eg. \f(CW\*(C`fuse\*(C'\fR) before you can use any FUSE filesystem.  This is necessary on Debian and derivatives.
.PP
On other distros, no special group is required.  It is not necessary on Fedora or Red Hat Enterprise Linux.
.SS "fusermount エラー: ""デバイスまたはリソースがビジーです"""
.IX Subsection "fusermount エラー: ""デバイスまたはリソースがビジーです"""
You can see this error when another process on the system jumps into the mountpoint you have just created, holding it open and preventing you from unmounting it.  The usual culprits are various GUI "indexing" programs.
.PP
The popular workaround for this problem is to retry the \f(CW\*(C`fusermount \-u\*(C'\fR command a few times until it works (\fBguestunmount\fR\|(1) does this for you). Unfortunately this isn\*(Aqt a reliable fix if (for example)  the mounted filesystem is particularly large and the intruding program particularly persistent.
.PP
A proper fix is to use a private mountpoint by creating a new mount namespace using the Linux\-specific \fBclone\fR\|(2)/\fBunshare\fR\|(2) flag \f(CW\*(C`CLONE_NEWNS\*(C'\fR.  Unfortunately at the moment this requires root and we would also probably need to add it as a feature to guestmount.
.SS "Race conditions possible when shutting down the connection"
.IX Subsection "Race conditions possible when shutting down the connection"
When \fBguestunmount\fR\|(1)/\fBfusermount\fR\|(1) exits, guestmount may still be running and cleaning up the mountpoint.  The disk image will not be fully finalized.
.PP
This means that scripts like the following have a nasty race condition:
.PP
.Vb 4
\& guestmount \-a disk.img \-i /mnt
\& # copy things into /mnt
\& guestunmount /mnt
\& # immediately try to use \*(Aqdisk.img\*(Aq ** UNSAFE **
.Ve
.PP
The solution is to use the \fI\-\-pid\-file\fR option to write the guestmount PID to a file, then after guestunmount spin waiting for this PID to exit.
.PP
.Vb 1
\& guestmount \-a disk.img \-i \-\-pid\-file guestmount.pid /mnt
\& 
\& # ...
\& # ...
\& 
\& # Save the PID of guestmount *before* calling guestunmount.
\& pid="$(cat guestmount.pid)"
\& 
\& # Unmount the filesystem.
\& guestunmount /mnt
\& 
\& timeout=10
\& 
\& count=$timeout
\& while kill \-0 "$pid" 2>/dev/null && [ $count \-gt 0 ]; do
\&     sleep 1
\&     ((count\-\-))
\& done
\& if [ $count \-eq 0 ]; then
\&     echo "$0: wait for guestmount to exit failed after $timeout seconds"
\&     exit 1
\& fi
\& 
\& # Now it is safe to use the disk image.
.Ve
.PP
Note that if you use the \f(CW\*(C`guestfs_mount_local\*(C'\fR API directly (see "MOUNT LOCAL" in \fBguestfs\fR\|(3)) then it is much easier to write a safe, race\-free program.
.SH オプション
.IX Header "オプション"
.IP "\fB\-a\fR IMAGE" 4
.IX Item "-a IMAGE"
.PD 0
.IP "\fB\-\-add\fR IMAGE" 4
.IX Item "--add IMAGE"
.PD
ブロックデバイスまたは仮想マシンイメージを追加します。
.Sp
ディスクイメージの形式は自動検知されます。 これを上書きして強制的に特定の形式を使用する場合、 \fI\-\-format=..\fR オプションを使用します。
.IP "\fB\-a\fR URI" 4
.IX Item "-a URI"
.PD 0
.IP "\fB\-\-add\fR URI" 4
.IX Item "--add URI"
.PD
リモートディスクを追加します。 "リモートストレージの追加" in \fBguestfish\fR\|(1) 参照。
.IP \fB\-\-blocksize=512\fR 4
.IX Item "--blocksize=512"
.PD 0
.IP \fB\-\-blocksize=4096\fR 4
.IX Item "--blocksize=4096"
.IP \fB\-\-blocksize\fR 4
.IX Item "--blocksize"
.PD
This parameter sets the sector size of the disk image.  It affects all explicitly added subsequent disks after this parameter.  Using \fI\-\-blocksize\fR with no argument switches the disk sector size to the default value which is usually 512 bytes.  See also "guestfs_add_drive_opts" in \fBguestfs\fR\|(3).
.IP "\fB\-c\fR URI" 4
.IX Item "-c URI"
.PD 0
.IP "\fB\-\-connect\fR URI" 4
.IX Item "--connect URI"
.PD
.Vb 1
\& I<\-d> オプションと同時に使用するとき、 これは使用する libvirt URI を指定します。 標準状態で標準の libvirt 接続を使用します。
.Ve
.IP "\fB\-d\fR LIBVIRT\-DOMAIN" 4
.IX Item "-d LIBVIRT-DOMAIN"
.PD 0
.IP "\fB\-\-domain\fR LIBVIRT\-DOMAIN" 4
.IX Item "--domain LIBVIRT-DOMAIN"
.PD
名前付き libvirt 仮想マシンからディスクを追加します。 \fI\-\-ro\fR オプションも使用されている場合、すべての libvirt 仮想マシンを使用できます。 しかしながら、書き込みモードでは、 停止状態の libvirt 仮想マシンのみ指定できます。
.Sp
名前の代わりに仮想マシンの UUID を使用できます。
.IP "\fB\-\-dir\-cache\-timeout\fR N" 4
.IX Item "--dir-cache-timeout N"
Set the readdir cache timeout to \fIN\fR seconds, the default being 60 seconds.  The readdir cache [actually, there are several semi\-independent caches] is populated after a \fBreaddir\fR\|(2) call with the stat and extended attributes of the files in the directory, in anticipation that they will be requested soon after.
.Sp
There is also a different attribute cache implemented by FUSE (see the FUSE option \fI\-o attr_timeout\fR), but the FUSE cache does not anticipate future requests, only cache existing ones.
.IP \fB\-\-echo\-keys\fR 4
.IX Item "--echo-keys"
キーやパスフレーズを入力するとき、通常 guestfish はエコーを無効化します。 そのため、入力内容を確認できません。 テンペスト攻撃の心配がなく、 部屋に誰も居なければ、 入力内容を確認するためにこのフラグを指定できます。
.IP \fB\-\-fd=\fRFD 4
.IX Item "--fd=FD"
Specify a pipe or eventfd file descriptor.  When the mountpoint is ready to be used, guestmount writes a single byte to this file descriptor.  This can be used in conjunction with \fI\-\-no\-fork\fR in order to run guestmount captive under another process.
.IP \fB\-\-format=raw|qcow2|..\fR 4
.IX Item "--format=raw|qcow2|.."
.PD 0
.IP \fB\-\-format\fR 4
.IX Item "--format"
.PD
\&\fI\-a\fR オプションは標準状態でディスクイメージの形式を自動検知します。 これを使用することにより、コマンドラインで後続の \fI\-a\fR オプションのディスク形式を強制的に指定できます。 引数なしで \fI\-\-format\fR を使用することにより、 後続の \fI\-a\fR オプションに対して自動検知に戻せます。
.Sp
仮想マシンのディスクイメージが信頼できない raw 形式である場合、 ディスク形式を指定するためにこのオプションを使用すべきです。 これにより、悪意のある仮想マシンにより起こり得る セキュリティ問題を回避できます (CVE\-2010\-3851)。 "\fBguestfs\fR\|(3)/guestfs_add_drive_opts" 参照。
.IP \fB\-\-fuse\-help\fR 4
.IX Item "--fuse-help"
Display help on special FUSE options (see \fI\-o\fR below).
.IP \fB\-\-help\fR 4
.IX Item "--help"
簡単なヘルプを表示して、終了します。
.IP \fB\-i\fR 4
.IX Item "-i"
.PD 0
.IP \fB\-\-inspector\fR 4
.IX Item "--inspector"
.PD
\&\fBvirt\-inspector\fR\|(1) コードを使用すると、 オペレーティングシステムを判定するためにディスクを検査します。 また、実際の仮想マシンにマウントすることと同じように ファイルシステムをマウントします。
.IP "\fB\-\-key\fR SELECTOR" 4
.IX Item "--key SELECTOR"
Specify a key for LUKS, to automatically open a LUKS device when using the inspection.

ChangeLog
commit 7264e4a4e059b775d9c28c60b589ddada94d98b8
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Sep 7 14:02:54 2021 +0100

    Version 1.44.2.

commit 1f0e713ecfdf9f7672a790e45256c613c6fda0ea
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Sep 7 16:59:56 2021 +0100

    builder: Suppress more bogus -fanalyzer warnings

commit 7f0ad952216879552e323a84544517e378609303
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Sep 7 16:30:28 2021 +0100

    rescue: Ignore -Wnull-dereference
    
    This appears somewhere in virt-rescue, unclear from where:
    
    make[2]: Entering directory '/home/rjones/d/libguestfs-1.44/rescue'
      CC       virt_rescue-suggest.o
      CCLD     virt-rescue
    rescue.c: In function 'main':
    ../common/utils/stringlists-utils.c:54:19: error: potential null pointer dereference [-Werror=null-dereference]
       54 |   for (r = 0; argv[r]; ++r)
          |                   ^
    ../common/utils/stringlists-utils.c:54:19: error: potential null pointer dereference [-Werror=null-dereference]
    lto1: all warnings being treated as errors

commit 46959189951df38fc6c64975fbaeab1afa4b1cc4
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Sep 7 15:53:07 2021 +0100

    daemon: Disable -Wstack-usage warning
    
    When using LTO, inlining something (I'm unclear what exactly) causes
    these errors:
    
    stubs-1.c: In function 'md_create_stub':
    stubs-1.c:1929:1: error: stack usage might be unbounded [-Werror=stack-usage=]
     1929 | md_create_stub (XDR *xdr_in)
          | ^
    stubs-0.c: In function 'mkfs_btrfs_stub':
    stubs-0.c:1984:1: error: stack usage might be unbounded [-Werror=stack-usage=]
     1984 | mkfs_btrfs_stub (XDR *xdr_in)
          | ^
    stubs-3.c: In function 'btrfs_device_delete_stub':
    stubs-3.c:477:1: error: stack usage might be unbounded [-Werror=stack-usage=]
      477 | btrfs_device_delete_stub (XDR *xdr_in)
          | ^
    stubs-3.c: In function 'btrfs_image_stub':
    stubs-3.c:694:1: error: stack usage might be unbounded [-Werror=stack-usage=]
      694 | btrfs_image_stub (XDR *xdr_in)
          | ^
    stubs-5.c: In function 'btrfs_device_add_stub':
    stubs-5.c:247:1: error: stack usage might be unbounded [-Werror=stack-usage=]
      247 | btrfs_device_add_stub (XDR *xdr_in)
          | ^
    
    Note this is only for the stable branch.

commit 704e84f213d4ae4ac4fcecc787985a886ff65e49
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Sep 7 14:56:56 2021 +0100

    daemon/xfs.c: Ignore bogus warnings from -fanalyzer
    
    Lots of variations of this.  It cannot happen because we check that
    section != NULL just before calling this code.
    
                    'parse_xfs_info': event 41
                      |
                      |../lib/guestfs-internal-all.h:50:21:
                      |   50 | #define STREQ(a,b) (strcmp((a),(b)) == 0)
                      |      |                     ^~~~~~~~~~~~~~~
                      |      |                     |
                      |      |                     (41) argument 1 ('section') from (35) could be NULL where non-null expected
    xfs.c:186:13: note: in expansion of macro 'STREQ'
                      |  186 |         if (STREQ (section, "meta-data")) {
                      |      |             ^~~~~
                      |
    
    Similarly this error does not appear possible:
    
    xfs.c:154:7: error: double-'free' of 'ret_9' [CWE-415] [-Werror=analyzer-double-free]
      154 |       free (section);
          |       ^
    
    See comment for why these suppressions are placed centrally.

commit a252ed5feb86dc9a458b1b9bd1bf924de3e32250
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Sep 7 14:36:18 2021 +0100

    lib/launch-libvirt.c: Fix type of outlen parameter of base64_decode_alloc_ctx
    
    In file included from launch-libvirt.c:49:
    launch-libvirt.c: In function 'add_secret':
    launch-libvirt.c:1956:54: error: pointer targets in passing argument 5 of 'base64_decode_alloc_ctx' differ in signedness [-Werror=pointer-sign]
     1956 |                               (char **) &secret_raw, &secret_raw_len)) {
          |                                                      ^~~~~~~~~~~~~~~
          |                                                      |
          |                                                      size_t * {aka long unsigned int *}
    ../gnulib/lib/base64.h:62:56: note: in definition of macro 'base64_decode_alloc'
       62 |         base64_decode_alloc_ctx (NULL, in, inlen, out, outlen)
          |                                                        ^~~~~~
    ../gnulib/lib/base64.h:56:57: note: expected 'idx_t *' {aka 'long int *'} but argument is of type 'size_t *' {aka 'long unsigned int *'}
       56 |                                      char **out, idx_t *outlen);
          |                                                  ~~~~~~~^~~~~~
    cc1: all warnings being treated as errors
    
    Note this patch is only required in the stable branch because
    of the gnulib update.

commit 64a11a274b2ca39584488ebb596b85080cd84890
Author: Richard W.M. Jones <rjones@redhat.com>
Date:   Tue Sep 7 14:30:59 2021 +0100


fuse/guestmount.c
/* guestmount - mount guests using libguestfs and FUSE
 * Copyright (C) 2009-2025 Red Hat Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <error.h>
#include <getopt.h>
#include <signal.h>
#include <limits.h>
#include <locale.h>
#include <libintl.h>

/* We're still using some of FUSE to handle command line options. */
#include <fuse.h>

#include "guestfs.h"

#include "ignore-value.h"
#include "getprogname.h"

#include "options.h"
#include "display-options.h"

static int write_pipe_fd (int fd);
static int write_pid_file (const char *pid_file, pid_t pid);

#ifndef HAVE_FUSE_OPT_ADD_OPT_ESCAPED
/* Copied from lib/fuse_opt.c and modified.
 * Copyright (C) 2001-2007  Miklos Szeredi <miklos@szeredi.hu>
 * This [function] can be distributed under the terms of the GNU LGPLv2.
 */
static void
fuse_opt_add_opt_escaped (char **opts, const char *opt)
{
  const unsigned oldlen = *opts ? strlen(*opts) : 0;
  char *d = realloc (*opts, oldlen + 1 + strlen(opt) * 2 + 1);

  if (!d)
    error (EXIT_FAILURE, errno, "realloc");

  *opts = d;
  if (oldlen) {
    d += oldlen;
    *d++ = ',';
  }

  for (; *opt; opt++) {
    if (*opt == ',' || *opt == '\\')
      *d++ = '\\';
    *d++ = *opt;
  }
  *d = '\0';
}
#endif

guestfs_h *g = NULL;
int read_only = 0;
int verbose = 0;
int inspector = 0;
int keys_from_stdin = 0;
int echo_keys = 0;
const char *libvirt_uri;
int in_guestfish = 0;
int in_virt_rescue = 0;

static void __attribute__((noreturn))
fuse_help (void)
{
  static struct fuse_operations null_operations;
  const char *tmp_argv[] = { getprogname (), "--help", NULL };
  fuse_main (2, (char **) tmp_argv, &null_operations, NULL);
  exit (EXIT_SUCCESS);
}

static void __attribute__((noreturn))
usage (int status)
{
  if (status != EXIT_SUCCESS)
    fprintf (stderr, _("Try ‘%s --help’ for more information.\n"),
             getprogname ());
  else {
    printf (_("%s: FUSE module for libguestfs\n"
              "%s lets you mount a virtual machine filesystem\n"
              "Copyright (C) 2009-2025 Red Hat Inc.\n"
              "Usage:\n"
              "  %s [--options] mountpoint\n"
              "Options:\n"
              "  -a|--add image       Add image\n"
              "  --blocksize[=512|4096]\n"
              "                       Set sector size of the disk for -a option\n"
              "  -c|--connect uri     Specify libvirt URI for -d option\n"
              "  --dir-cache-timeout  Set readdir cache timeout (default 5 sec)\n"
              "  -d|--domain guest    Add disks from libvirt guest\n"
              "  --echo-keys          Don't turn off echo for passphrases\n"
              "  --fd=FD              Write to pipe FD when mountpoint is ready\n"
              "  --format[=raw|..]    Force disk format for -a option\n"
              "  --fuse-help          Display extra FUSE options\n"
              "  -i|--inspector       Automatically mount filesystems\n"
              "  --help               Display help message and exit\n"
              "  --key selector       Specify a LUKS key\n"
              "  --keys-from-stdin    Read passphrases from stdin\n"
              "  -m|--mount dev[:mnt[:opts[:fstype]] Mount dev on mnt (if omitted, /)\n"
              "  --no-fork            Don't daemonize\n"
              "  -n|--no-sync         Don't autosync\n"
              "  -o|--option opt      Pass extra option to FUSE\n"
              "  --pid-file filename  Write PID to filename\n"
              "  -r|--ro              Mount read-only\n"
              "  --selinux            For backwards compat only, does nothing\n"
              "  -v|--verbose         Verbose messages\n"
              "  -V|--version         Display version and exit\n"
              "  -w|--rw              Mount read-write\n"
              "  -x|--trace           Trace guestfs API calls\n"
              ),
            getprogname (), getprogname (),
            getprogname ());
  }
  exit (status);
}

int
main (int argc, char *argv[])
{
  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEBASEDIR);
  textdomain (PACKAGE);

  parse_config ();

  enum { HELP_OPTION = CHAR_MAX + 1 };

  /* The command line arguments are broadly compatible with (a subset
   * of) guestfish.  Thus we have to deal mainly with -a, -m and --ro.
   */
  static const char options[] = "a:c:d:im:no:rvVwx";
  static const struct option long_options[] = {
    { "add", 1, 0, 'a' },
    { "blocksize", 2, 0, 0 },
    { "connect", 1, 0, 'c' },
    { "dir-cache-timeout", 1, 0, 0 },
    { "domain", 1, 0, 'd' },
    { "echo-keys", 0, 0, 0 },
    { "fd", 1, 0, 0 },
    { "format", 2, 0, 0 },
    { "fuse-help", 0, 0, 0 },
    { "help", 0, 0, HELP_OPTION },
    { "inspector", 0, 0, 'i' },
    { "key", 1, 0, 0 },
    { "keys-from-stdin", 0, 0, 0 },
    { "live", 0, 0, 0 },
    { "long-options", 0, 0, 0 },
    { "mount", 1, 0, 'm' },
    { "no-fork", 0, 0, 0 },
    { "no-sync", 0, 0, 'n' },
    { "option", 1, 0, 'o' },
    { "pid-file", 1, 0, 0 },
    { "ro", 0, 0, 'r' },
    { "rw", 0, 0, 'w' },
    { "selinux", 0, 0, 0 },
    { "short-options", 0, 0, 0 },
    { "trace", 0, 0, 'x' },
    { "verbose", 0, 0, 'v' },
    { "version", 0, 0, 'V' },
    { 0, 0, 0, 0 }
  };

  struct drv *drvs = NULL;
  struct mp *mps = NULL;
  struct mp *mp;
  char *p;
  const char *format = NULL;
  bool format_consumed = true;
  int blocksize = 0;
  bool blocksize_consumed = true;
  int c, r;
  int option_index;
  struct key_store *ks = NULL;

  int debug_calls = 0;
  int dir_cache_timeout = -1;
  int do_fork = 1;
  char *fuse_options = NULL;
  char *pid_file = NULL;
  int pipe_fd = -1;

  struct guestfs_mount_local_argv optargs;

  /* LC_ALL=C is required so we can parse error messages. */
  setenv ("LC_ALL", "C", 1);

  sigaction (SIGPIPE, &(struct sigaction){ .sa_handler = SIG_IGN, .sa_flags = SA_RESTART }, NULL);

  g = guestfs_create ();
  if (g == NULL)
    error (EXIT_FAILURE, errno, "guestfs_create");

  for (;;) {
    c = getopt_long (argc, argv, options, long_options, &option_index);
    if (c == -1) break;

    switch (c) {
    case 0:			/* options which are long only */
      if (STREQ (long_options[option_index].name, "long-options"))
        display_long_options (long_options);
      else if (STREQ (long_options[option_index].name, "short-options"))
        display_short_options (options);
      else if (STREQ (long_options[option_index].name, "dir-cache-timeout"))
        dir_cache_timeout = atoi (optarg);
      else if (STREQ (long_options[option_index].name, "fuse-help"))
        fuse_help ();
      else if (STREQ (long_options[option_index].name, "selinux")) {
        /* nothing */
      } else if (STREQ (long_options[option_index].name, "format")) {
        OPTION_format;
      } else if (STREQ (long_options[option_index].name, "blocksize")) {
        OPTION_blocksize;
      } else if (STREQ (long_options[option_index].name, "keys-from-stdin")) {
        keys_from_stdin = 1;
      } else if (STREQ (long_options[option_index].name, "echo-keys")) {
        echo_keys = 1;
      } else if (STREQ (long_options[option_index].name, "live")) {
        error (EXIT_FAILURE, 0,
               _("libguestfs live support was removed in libguestfs 1.48"));
      } else if (STREQ (long_options[option_index].name, "pid-file")) {
        pid_file = optarg;
      } else if (STREQ (long_options[option_index].name, "no-fork")) {
        do_fork = 0;
      } else if (STREQ (long_options[option_index].name, "fd")) {
        if (sscanf (optarg, "%d", &pipe_fd) != 1 || pipe_fd < 0)
          error (EXIT_FAILURE, 0,
                 _("unable to parse --fd option value: %s"), optarg);
      } else if (STREQ (long_options[option_index].name, "key")) {
        OPTION_key;
      } else
        error (EXIT_FAILURE, 0,
               _("unknown long option: %s (%d)"),
               long_options[option_index].name, option_index);
      break;

    case 'a':
      OPTION_a;
      break;

    case 'c':
      OPTION_c;
      break;

    case 'd':
      OPTION_d;
      break;

    case 'i':
      OPTION_i;
      break;

    case 'm':
      OPTION_m;
      break;

    case 'n':
      OPTION_n;
      break;

    case 'o':
      fuse_opt_add_opt_escaped (&fuse_options, optarg);
      break;

    case 'r':
      OPTION_r;
      break;

    case 'v':
      OPTION_v;
      break;

    case 'V':
      OPTION_V;
      break;

    case 'w':
      OPTION_w;
      break;

    case 'x':
      OPTION_x;
      debug_calls = 1;
      do_fork = 0;
      break;

    case HELP_OPTION:
      usage (EXIT_SUCCESS);

    default:
      usage (EXIT_FAILURE);
    }
  }

  CHECK_OPTION_format_consumed;
  CHECK_OPTION_blocksize_consumed;

  /* Check we have the right options. */
  if (drvs == NULL) {
    fprintf (stderr, _("%s: error: you must specify at least one -a or -d option.\n"),
             getprogname ());
    usage (EXIT_FAILURE);
  }
  if (!(mps || inspector)) {
    fprintf (stderr, _("%s: error: you must specify either -i at least one -m option.\n"),
             getprogname ());
    usage (EXIT_FAILURE);
  }

  /* We'd better have a mountpoint. */
  if (optind+1 != argc)
    error (EXIT_FAILURE, 0,
           _("you must specify a mountpoint in the host filesystem"));

  /* If we're forking, we can't use the recovery process. */
  if (guestfs_set_recovery_proc (g, !do_fork) == -1)
    exit (EXIT_FAILURE);

  /* Do the guest drives and mountpoints. */
  add_drives (drvs);

  if (key_store_requires_network (ks) && guestfs_set_network (g, 1) == -1)
    exit (EXIT_FAILURE);

  if (guestfs_launch (g) == -1)
    exit (EXIT_FAILURE);
  if (inspector)
    inspect_mount ();
  mount_mps (mps);

  free_drives (drvs);
  free_mps (mps);
  free_key_store (ks);

  /* FUSE example does this, not clear if it's necessary, but ... */
  if (guestfs_umask (g, 0) == -1)
    exit (EXIT_FAILURE);

  optargs.bitmask = 0;
  if (read_only) {
    optargs.bitmask |= GUESTFS_MOUNT_LOCAL_READONLY_BITMASK;
    optargs.readonly = 1;
  }
  if (debug_calls) {
    optargs.bitmask |= GUESTFS_MOUNT_LOCAL_DEBUGCALLS_BITMASK;
    optargs.debugcalls = 1;
  }
  if (dir_cache_timeout > 0) {
    optargs.bitmask |= GUESTFS_MOUNT_LOCAL_CACHETIMEOUT_BITMASK;
    optargs.cachetimeout = dir_cache_timeout;
  }
  if (fuse_options != NULL) {
    optargs.bitmask |= GUESTFS_MOUNT_LOCAL_OPTIONS_BITMASK;
    optargs.options = fuse_options;
  }

  if (guestfs_mount_local_argv (g, argv[optind], &optargs) == -1)
    exit (EXIT_FAILURE);

  /* Daemonize. */
  if (do_fork) {
    pid_t pid;
    int fd;

    pid = fork ();
    if (pid == -1)
      error (EXIT_FAILURE, errno, "fork");

    if (pid != 0) {             /* parent */
      if (write_pid_file (pid_file, pid) == -1)
        _exit (EXIT_FAILURE);
      if (write_pipe_fd (pipe_fd) == -1)
        _exit (EXIT_FAILURE);

      _exit (EXIT_SUCCESS);
    }

    /* Emulate what old fuse_daemonize used to do. */
    if (setsid () == -1)
      error (EXIT_FAILURE, errno, "setsid");

    ignore_value (chdir ("/"));

    fd = open ("/dev/null", O_RDWR);
    if (fd >= 0) {
      dup2 (fd, 0);
      dup2 (fd, 1);
      dup2 (fd, 2);
      if (fd > 2)
        close (fd);
    }
  }
  else {
    /* not forking, write PID file and pipe FD anyway */
    if (write_pid_file (pid_file, getpid ()) == -1)
      exit (EXIT_FAILURE);
    if (write_pipe_fd (pipe_fd) == -1)
      exit (EXIT_FAILURE);
  }

  /* At the last minute, remove the libguestfs error handler.  In code
   * above this point, the default error handler has been used which
   * sends all errors to stderr.  From now on, the FUSE code will
   * convert errors into error codes (errnos) when appropriate.
   */
  guestfs_push_error_handler (g, NULL, NULL);

  /* Main loop. */
  r = guestfs_mount_local_run (g);

  guestfs_pop_error_handler (g);

  /* Cleanup. */
  if (guestfs_shutdown (g) == -1)
    r = -1;
  guestfs_close (g);

  /* Don't delete PID file until the cleanup has been completed. */
  if (pid_file)
    unlink (pid_file);

  exit (r == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
}

static int
write_pid_file (const char *pid_file, pid_t pid)
{
  FILE *fp;

  if (pid_file == NULL)
    return 0;

  fp = fopen (pid_file, "w");
  if (fp == NULL) {
  error:
    perror (pid_file);
    return -1;
  }

  if (fprintf (fp, "%d\n", pid) == -1) {
    fclose (fp);
    goto error;
  }

  if (fclose (fp) == -1)
    goto error;

  return 0;
}

static int
write_pipe_fd (int fd)
{
  char c = 0;

  if (fd < 0)
    return 0;

  if (write (fd, &c, 1) != 1) {
    perror ("write (--fd option)");
    return -1;
  }

  if (close (fd) == -1) {
    perror ("close (--fd option)");
    return -1;
  }

  return 0;
}
